### PR TITLE
Raise nicer exceptions for bad generator and filters in comprehensions

### PR DESCRIPTION
**Pull request by bjorng**
This pull request introduce a nicer `{bad_generator,Generator}` exception when a generator for list or binary comprehension has the wrong type, and a `{bad_filter,Filter}` exception when a filter does not evaluate to a boolean. This PR also fixes some comprehension bugs and refactor and simplifies compiler code.

### CODE REVIEW

File: lib/compiler/src/beam_ssa_bc_size.erl [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+join_bs_1([], _Bigger, Smaller) -> Smaller.
+
+%%%
+%%% Turn the result of the traversal of the SSA code into an expression tree.
+%%%
+
+make_expr_tree({{call,Alloc0},GenAnno}) ->
+    {Alloc1,Annos} = make_expr_tree_list(Alloc0, none, none, [GenAnno]),
+    Alloc2 = opt_expr(Alloc1),
+    Alloc = case divisible_by_eight(Alloc2) of
```
Comment (jhogberg) [inline lib/compiler/src/beam_ssa_bc_size.erl:372]: I think you should add a comment or break this out into its own function. At a first glance it's not obvious we're rounding to the next byte.

File: lib/compiler/src/beam_ssa_type.erl [hunk 2]
```diff
@@ -1349,6 +1361,10 @@ eval_type_test_bif(I, Op, Types) ->
             I
     end.
 
+t_is_number(#t_integer{}) -> true;
```
Comment (jhogberg) [inline lib/compiler/src/beam_ssa_type.erl:1364]: I'd prefer this were added to `beam_types` where we already have similar functions. `is_numerical_type` would fit right in.

**Review by jhogberg — APPROVED** [review]
Looks good to me aside from a few minor nits.

File: lib/compiler/src/beam_ssa_bc_size.erl [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+    end;
+opt_expr_1({bif,'div'}=Op, [Numerator,#b_literal{val=Denominator}]=Args) ->
+    try
+        opt_expr_div(Numerator, Denominator)
+    catch
+        throw:not_possible ->
+            case Denominator band (Denominator - 1) of
+                0 ->
+                    %% The denominator is a power of two.
+                    {{bif,'bsr'},[Numerator,#b_literal{val=log2(Denominator)}]};
```
[the remark this answers is not in the document]
Comment (bjorng) [reply lib/compiler/src/beam_ssa_bc_size.erl:447]: I'll use `round(math:log2(Denominator))`.

**Comment by bjorng** [comment]
I have updated this PR to also raise `{bad_filter,Filter}` exceptions for bad filters.

**Pull request merged**